server/src/main/java/io/druid/server/QueryResource.java

```diff
@@ -220,8 +222,17 @@ public void write(OutputStream outputStream) throws IOException, WebApplicationE
                 },
                 contentType
             )
-            .header("X-Druid-Query-Id", queryId)
-            .header("X-Druid-Response-Context", jsonMapper.writeValueAsString(responseContext))
+            .header("X-Druid-Query-Id", queryId);
+
+        //Limit the response-context header, see https://github.com/druid-io/druid/issues/2331
+        String responseCtxString = jsonMapper.writeValueAsString(responseContext);
+        if (responseCtxString.length() > RESPONSE_CTX_HEADER_LEN_LIMIT) {
+          log.warn("Response Context truncated for id [%s] . Full context is [%s].", queryId, responseCtxString);
+          responseCtxString = responseCtxString.substring(0, RESPONSE_CTX_HEADER_LEN_LIMIT);
```

Comment:
the only reason I can think of doing that is that it might help debugging without having access to the server logs. If your truncated context has a ton of the same kind of object in it then that would give you some hints about what is going on.